### PR TITLE
Add created and updated on the users events

### LINKED ISSUE #106: No order property in entities
Every entity (special with no sequence ids) should have created time or smth else to sort them in lists or collections.

### PR DESCRIPTION
**Pull request by warguns**
Fixes #106 

- Created a Domain value object and Exception for the DateTime
- Created Doctrine Datetime Type for easy manipulation
- Added created_at and updated_at on user's events and projections
- Added the two fields on the users' table
- Change UserSignedIn factory method in favor of __construct

### CODE REVIEW

File: src/Domain/User/User.php [hunk 1]
```diff
@@ -60,6 +67,8 @@ protected function applyUserWasCreated(UserWasCreated $event): void
 
         $this->setEmail($event->credentials->email);
         $this->setHashedPassword($event->credentials->password);
+        $this->setCreatedAt($event->createdAt);
+        $this->setUpdatedAt($event->updatedAt);
```
Comment (jorge07) [inline src/Domain/User/User.php]: Remove updatedAt here

File: src/Domain/User/User.php [hunk 2]
```diff
@@ -82,6 +92,26 @@ private function setHashedPassword(HashedPassword $hashedPassword): void
...
+    }
+
+    public function createdAt(): string
+    {
+        return $this->createdAt->toString();
+    }
+
+    public function updatedAt(): string
+    {
+        return $this->updatedAt->toString();
```
Comment (jorge07) [inline src/Domain/User/User.php]: This value must be nullable for never updated aggregates

File: src/Domain/User/User.php [hunk 3]
```diff
@@ -105,4 +135,10 @@ public function getAggregateRootId(): string
 
     /** @var HashedPassword */
     private $hashedPassword;
+
+    /** @var DateTime */
+    private $createdAt;
+
+    /** @var DateTime */
```
Comment (jorge07) [inline src/Domain/User/User.php]: ?DateTime

File: src/Infrastructure/Share/Doctrine/DateTimeType.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+use App\Domain\Shared\Exception\DateTimeException;
+use App\Domain\Shared\ValueObject\DateTime;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Types\ConversionException;
+use Doctrine\DBAL\Types\Type;
+
+class DateTimeType extends Type
+{
+    const NAME = 'domain_datetime_immutable';
```
Comment (jorge07) [inline src/Infrastructure/Share/Doctrine/DateTimeType.php]: Remove domain prefix

File: src/Infrastructure/Share/Doctrine/DateTimeType.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            );
+        }
+
+        return $dateTime;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function requiresSQLCommentHint(AbstractPlatform $platform)
```
Comment (jorge07) [inline src/Infrastructure/Share/Doctrine/DateTimeType.php]: Should this method live here?

File: src/Infrastructure/Share/Migrations/Version20190415193428.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace DoctrineMigrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+class Version20190415193428 extends AbstractMigration
```
Comment (jorge07) [inline src/Infrastructure/Share/Migrations/Version20190415193428.php]: You'll need to recreate this after the changes

File: tests/Domain/User/UserTest.php [hunk 7]
```diff
@@ -130,4 +130,33 @@ public function isUnique(Email $email): bool
...
+            Uuid::uuid4(),
+            new Credentials(
+                Email::fromString($emailString),
+                HashedPassword::encode('password')
+            ),
+            $this
+        );
+
+        self::assertNotNull($user->createdAt());
+        self::assertNotNull($user->updatedAt());
```
Comment (jorge07) [inline tests/Domain/User/UserTest.php]: Must be null here

File: src/Domain/User/Event/UserSignedIn.php [hunk 8]
```diff
@@ -44,6 +34,12 @@ public function serialize(): array
         ];
     }
 
+    public function __construct(UuidInterface $uuid, Email $email)
```
[the remark this answers is not in the document]
Comment (warguns) [reply src/Domain/User/Event/UserSignedIn.php]: Neither UserEmailChanged, nor UserWasCreated DTO's are using a factory method to autoconstruct itself, also in that factory wasn't doing any complex construction. So in order to maintain coherence between the DTO's I changed it. Maybe we can create factory methods for all the DTO's, what do you think?

File: src/Domain/Shared/ValueObject/DateTime.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+        return $self;
+    }
+
+    public function toString(): string
+    {
+        return $this->dateTime->format(self::FORMAT);
+    }
+
+    public function toFormat(string $format): string
```
[the remark this answers is not in the document]
Comment (warguns) [reply src/Domain/Shared/ValueObject/DateTime.php]: Hmmm... that toFormat was only used on the Doctrine Type, so lastly I replaced it with the toNative method returning the original DateTimeImmutable.

[the remark this answers is not in the document]
Comment (jorge07) [reply src/Domain/Shared/ValueObject/DateTime.php]: 👍

File: src/Domain/User/User.php [hunk 10]
```diff
@@ -82,6 +91,26 @@ private function setHashedPassword(HashedPassword $hashedPassword): void
         $this->hashedPassword = $hashedPassword;
     }
 
+    private function setCreatedAt(DateTime $createdAt): void
+    {
+        $this->createdAt = $createdAt;
+    }
+
+    private function setUpdatedAt(DateTime $updatedAt): void
```
Comment (jorge07) [inline src/Domain/User/User.php]: Thinking about remove all this setters. Also SetEmail and SetHashedPassword. Thoughts?

File: src/Domain/User/User.php (hunk 3 above)
Comment (jorge07) [reply src/Domain/User/User.php]: @warguns this needs to be nullable

File: src/Infrastructure/Share/Migrations/Version20190416102503.php [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20190416102503 extends AbstractMigration
+{
+    public function up(Schema $schema): void
+    {
+        $this->abortIf('mysql' !== $this->connection->getDatabasePlatform()->getName(), 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE users ADD created_at DATETIME NOT NULL COMMENT \'(DC2Type:datetime_immutable)\', ADD updated_at DATETIME DEFAULT NULL COMMENT \'(DC2Type:datetime_immutable)\'');
```
Comment (jorge07) [inline src/Infrastructure/Share/Migrations/Version20190416102503.php]: 👍

**Comment by jorge07** [comment]
@warguns almost done. Just few notes

File: src/Domain/User/User.php (hunk 10 above)
Comment (warguns) [reply src/Domain/User/User.php]: I agree with you, if we manage all the changes using the apply methods it does not have sense to maintain the setters.

**Comment by jorge07** [comment]
Merged. Thanks @warguns

**Pull request merged**